### PR TITLE
Unification: A type-checking fix in imitation + an error locating fix

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** bug fix

Overlay for Equations which apparently works "better": mattam82/Coq-Equations#316

We fix a pending typing issue in `solve_simple_eqn` when imitation stops to be possible and an equation is postponed (this was observed while working on a complex unification problem).

We also avoid raising an error (in `recheck_application`) which bypasses the exception handler of `solve_simple_eqn`, resulting in a lack of location. It is actually not clear if this `recheck_application` unlocated error can really pop up. It did in the example I was working on due to the previous typing issue.

### CODE REVIEW

**Comment by SkySkimmer** [comment]
>We fix a pending typing issue in solve_simple_eqn when imitation stops to be possible and an equation is postponed (this was observed while working on a complex unification problem).

Is adding a test possible?

**Comment by herbelin** [comment]
> Is adding a test possible?

It is difficult. I have a 30-line definition with a unification problem which errors and highlights the whole definition without the fix and which returns another, though highlighted, error unification with the fix. 

We coult test the different error messages if really necessary, but the risk is that this file becomes difficult to maintain when e.g. small changes here and there are done. The whole error message, after the fix, is itself several dozens of lines long and involves a lot of evars with their instances.

I can send you the file if you want to try to extract a reasonable test out of it though.

**Comment by herbelin** [comment]
This PR is ready. @maximedenes, as a pretyper maintainer, would you assign for this fix of a ill-typed assignment in the imitation phase of unification?

**Comment by ejgallego** [comment]
@maximedenes this is tagged 8.12.1 , do you think it is ready ?

**Comment by maximedenes** [comment]
@ejgallego AFAICT, this PR lacks a review. I'm unfortunately not competent on this part of the code, and it is almost completely undocumented, so it's hard to acquire knowledge. So as an assignee, all I can do is it to call for help so that somebody competent reviews the code. Maybe @mattam82 or @gares?

**Comment by maximedenes** [comment]
Also, the equations overlay seems to be waiting for feedback from @mattam82.

**Comment by ejgallego** [comment]
Indeed we need feedback from @mattam82 too as this would break inter-release compat for 8.12.1 (cc: @Zimmi48 )

**Comment by herbelin** [comment]
@maximedenes:

> and it is almost completely undocumented, so it's hard to acquire knowledge.

Shortly, the `invert_definition` part is about solving "arbitrary" equations of the form `Γ ⊢ ?x@{ai:=ui} == t` where `a1:T1, ..., an:Tn ⊢ ?x : U`.

Typical such problems are `y:nat ⊢ ?x@{a:=y} = S  y` for `a : nat ⊢ ?x : nat`, for which the unique (canonical) solution is `?x:=a`.

The PR is about the case when `t` is a variable, i.e. when the problem has the form `yj:Vi ⊢?x{ai:=ui} == yj₀`. In this case, because `?x` is in context `a1:T1, ..., an:Tn`, it is not allowed to be `yj₀` itself, so, the only hope to solve the equation is that `?x` is instantiated by something which extract `yj₀` from one the `ui`.

A typical such case is when one of the `ui`, say `ui₀` is precisely `yj₀`, in which case setting `?x:=ai₀` is enough to ensure that `?x{ai:=ui} == yj₀`. This corresponds to the case `ProjectVar` of the function `find_solution_type` removed by the PR.

Another such typical case is when one of the `ui`, say `ui₀`, is an evar applied to `yj₀`, say `?y@{b:=yj₀}`. In this case, it is enough to define both `?x:=ai₀` and `?y:=b` to ensure that `?x{ai:=ui} == yj₀`. This corresponds to the case `ProjectEvar` of the function `find_solution_type` removed by the PR.

There might be several such solutions, this is why we build a list of such `sols`, made of `ProjectVar` and `ProjectEvar`. (There are other possible solutions, but this part of the code makes an approximation and drops the other forms of solutions.)

When there is a unique solution, we assert that this is _the_ solution (even though the uniqueness is only wrt the kind of inversions we do - in the longer term, we would need to be more cautious about not dropping other forms of solutions).

When there are several solutions, we postpone (a typical case of postponment is `y:nat ⊢ ?x@{a:=y,a':=y} = y` which has two solutions.

For some reason which I attributed to naivety at that time, I used in the past a typing function for `?x@{ai:=ui}` which was typing the first solution found for `?x` (then instantiating it with `ai:=ui`) rather than typing `?x@{ai:=ui}` directly. If the solutions are well-typed, this should not make a difference and I don't see here how they could be ill-typed.

Additionally, this typing function was knowledgeably bugged (as the comment says). I see no reason to keep it, so I use instead the obvious function typing `?x@{ai:=ui}` (it returns a type in context `Γ`).

The `of_alias` is irrelevant, it is the trivial embedding of the datatype of `Rel` or `Var` to a `constr`.


The `recheck_applications` application part is about calling an exception `cannot_unify` which bypasses the handler of `solve_simple_eqn`. But `solve_simple_eqn` is rather supposed to return either a `Success` or a `UnifFailure` so that backtracking from `solve_simple_eqn` is possible. I changed the `cannot_unify` to an error catchable by `solve_simple_eqn`. Additionally, this avoids highlighting the whole expression since `cannot_unify` was bypassing also the localisation of the error. (In the case of the example above, this localisation was done in `inh_conv_coerce_to_gen` thanks to the call to `error_actual_type`.)

Tell me if you'd like more comments.

**Comment by gares** [comment]
To my understanding the fix is reasonable, but I don't think it is reasonable to change unification between point releases (I did not check the overlay, but if there is one it means something changed).
If the error message part can be separated (its unclear from your description) then that part can go in a point release for sure.

**Comment by herbelin** [comment]
I don't think there is a urgency. Targetting 8.13 would probably be good enough.

Regarding the overlay, don't hesitate looking at it. It seems to be a clear fix: a generic tactic that was not working correctly and which was needing an extra manual tactic invocation.

**Comment by Zimmi48** [comment]
Thanks for the ping, @ejgallego. My analysis of the situation is:
- this is not a regression fix,
- this is a change to an unspecified part of the system,
- it has shown to have some (minor) compatibility impact already.

So it really does not belong to a patch-level release.

**Comment by Zimmi48** [comment]
There's a good amount of useful knowledge in @herbelin's comment. I wonder how to make this writing more permanent. Could it be transformed into comments to add to the code? Or should it be added to `dev/doc/unification.txt`?

**Comment by gares** [comment]
> Typical such problems are `y:nat ⊢ ?x@{a:=y} = S y` for `a : nat ⊢ ?x : nat`, for which the unique (canonical) solution is `?x:=a`.

If you do so, I believe here the solution should be `?x:= S a`

**Comment by maximedenes** [comment]
> There's a good amount of useful knowledge in @herbelin's comment. I wonder how to make this writing more permanent. Could it be transformed into comments to add to the code? Or should it be added to `dev/doc/unification.txt`?

In the current state, it would be good to add it to the code, as it is always easier to maintain in sync. Once the code is fully commented, we should be able to extract a higher level spec (to be integrated in the manual).

**Comment by herbelin** [comment]
> In the current state, it would be good to add it to the code, as it is always easier to maintain in sync. Once the code is fully commented, we should be able to extract a higher level spec (to be integrated in the manual).

For the current code, do the way you think it is useful.

In any case, it is important that the Coq dev team develops more and more expertise about the unification algorithm.

Also, the irregularities of the current code should be smoothened before documenting the code (i.e. the local ad hoc strategies which prevent obtaining a nice high level specification should be removed - e.g. failing too early rather than postponing, or losing solutions too early).  What we should ideally target is an extension of Matthieu's and Beta's paper which integrates all the ideas developed in the current implementation + new unification strategies which are very important (for instance, equations such as `?x@{a:=(y,z)} == y` are common trivially solvable equations and it is a misery that this is still not implemented).

**Comment by herbelin** [comment]
> > Typical such problems are `y:nat ⊢ ?x@{a:=y} = S y` for `a : nat ⊢ ?x : nat`, for which the unique (canonical) solution is `?x:=a`.
> 
> If you do so, I believe here the solution should be `?x:= S a`

Indeed! Let me fix my comment.

**Comment by maximedenes** [comment]
> Also, the irregularities of the current code should be smoothened before documenting the code

I guess that's where I disagree. I strongly believe we should document it first, so that we can onboard more people and then improve the situation with their help. For example, documentation would make it easier for people like me to review unification-related PRs.

**Comment by herbelin** [comment]
> Also, the irregularities of the current code should be smoothened before documenting the code

Sorry, there is a misunderstanding. I did not mean that I don't want to add comments to the code (on the contrary, as soon as we explain at the same time what is wrong in the code). I meant that I don't think it makes sense to document in the reference manual the defects of the current code of unification.

Of course, we need to onboard more people and then improve the situation with their help. That was the point of my previous sentence.

**Comment by Zimmi48** [comment]
Yes, explaining in the comments what is wrong with the current code as well is even better, because this is part of the useful knowledge. As for documenting this in the reference manual, it doesn't have to be as detailed as what is put in the code, but if some strange behavior comes up often in questions from users, then it is useful to also add it to the reference manual even if to say that this is some behavior that we are not satisfied with and intend to change.

**Comment by herbelin** [comment]
As far as I can recap, this PR is ready.

**Review by gares — APPROVED** [review]

**Comment by gares** [comment]
This looks good to me. IMO we should run a bench, but the PR probably predates the new infrastructure (I don't see the button to run the bench). Dear @coq/ci-maintainers would a rebase fix this?

**Comment by gares** [comment]
@herbelin would you mind rebasing? it will not hurt

**Comment by gares** [comment]
Great, bench at https://gitlab.com/coq/coq/-/jobs/851267101

**Comment by gares** [comment]
Bench is looking good. The overlay for Equations seems not necessary/spurious. Can you remove it @herbelin ?

**Comment by herbelin** [comment]
> The overlay for Equations seems not necessary/spurious. Can you remove it @herbelin ?

It was made last August and got outdated. I rebased it.

**Comment by gares** [comment]
@coqbot merge now

**Pull request merged**
